src/lib/child_processes/termination.ml

```diff
@@ -40,16 +40,25 @@ let get_signal_cause_opt =
 let get_child_data (t : t) child_pid = Pid.Table.find t child_pid
 
 let check_terminated_child (t : t) child_pid logger =
-  if Pid.Table.mem t child_pid then
-    let data = Pid.Table.find_exn t child_pid in
-    if not data.termination_expected then (
-      [%log error]
-        "Child process of kind $process_kind with pid $child_pid has terminated"
-        ~metadata:
-          [ ("child_pid", `Int (Pid.to_int child_pid))
-          ; ("process_kind", `String (show_process_kind data.kind)) ] ;
-      Core_kernel.exit 99 )
+  match get_child_data t child_pid with
+  | None ->
+      (* not a process of interest *)
+      ()
+  | Some data ->
+      if not data.termination_expected then (
+        let kind = show_process_kind data.kind in
+        [%log error]
+          "Child process of kind $process_kind with pid $child_pid has \
+           unexpectedly terminated"
+          ~metadata:
+            [ ("child_pid", `Int (Pid.to_int child_pid))
+            ; ("process_kind", `String kind) ] ;
+        failwithf "Child process of kind %s has unexpectedly terminated" kind
```

Comment:
Does this change behavior? As in, won't a `failwith` trigger graceful shutdown, where as an `exit` won't? I'm worried that this exit was intentionally, and that graceful shutdown might fail in this condition.

Reply:
I see now that you did mention in the description that this changes behavior. I'm still mildly fearful of this change, just given that you haven't (and couldn't) tested the libp2p child process dying, and that there are other ways that processes can die.

Reply:
I wrote the original code; the `exit` was written without any contemplation of graceful shutdown, or not. Don't we want to at least attempt the graceful shutdown?

A `libp2p` termination should not result in the `failwith` here, because the process is registered with `termination_expected = true`, from the code in #8415.